File: tests/src/test/java/messages/QuoteTest.java
package messages;

import java.util.UUID;

import org.agrona.DirectBuffer;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import static org.junit.jupiter.api.Assertions.assertEquals;
import org.junit.jupiter.api.Test;

import agrona.messages.MessageHeaderDecoder;
import agrona.messages.QuoteDecoder;
import sbe.SbeEncoder;

public class QuoteTest {
    private static final Logger log = LogManager.getLogger(QuoteTest.class);

    // Define quote data
    double amount;
    String currency;
    String side;
    String symbol;
    String transactTime;
    String quoteID;     
    String quoteRequestID;
    double fxRate;
    String clientID;
    double secondaryAmount; 

    @Test
    public void testSbeQuote() {
        log.info("In testSbeQuote");
        System.out.println("Test Working Directory: " + System.getProperty("user.dir"));

        // Create a random GUID
        UUID uuid = UUID.randomUUID();
        String guid = uuid.toString().substring(0, 8).toUpperCase();

        // Populate quote data
        amount = 1000;
        currency = "USD";
        side = "BUY";
        symbol = "EURUSD";
        transactTime = "20240101-00:00:00.000";
        quoteID = guid;     
        quoteRequestID = guid;
        fxRate = 1.23456;
        clientID = "TEST";
        secondaryAmount = 810.01; 

        // Initialize the encoder
        SbeEncoder sbeEncoder = new SbeEncoder();

        // Define the data to encode
        DirectBuffer dataBuffer = sbeEncoder.encodeQuote(amount, currency, side, symbol, transactTime,
                                                    quoteID, quoteRequestID, fxRate, secondaryAmount, clientID);

        // Decode the message
        QuoteDecoder decoder = new QuoteDecoder();
        decoder.wrap(dataBuffer, MessageHeaderDecoder.ENCODED_LENGTH, QuoteDecoder.BLOCK_LENGTH, QuoteDecoder.SCHEMA_VERSION);

        // Verify the decoded message
        assertEquals((long) (amount * 100), decoder.amount().mantissa());
        assertEquals(-2, decoder.amount().exponent());
        assertEquals(currency, decoder.currency());
        assertEquals((long) (secondaryAmount * 100), decoder.secondaryAmount().mantissa());
        assertEquals(-2, decoder.secondaryAmount().exponent());
        assertEquals(side, decoder.side());
        assertEquals(symbol, decoder.symbol());
        assertEquals(transactTime, decoder.transactTime());
        assertEquals(quoteRequestID, decoder.quoteRequestID());
        assertEquals(quoteID, decoder.quoteID());
        assertEquals((long) (fxRate * 100000), decoder.fxRate().mantissa());
        assertEquals(-5, decoder.fxRate().exponent());
        assertEquals(clientID, decoder.clientID());

        log.info("Received and verified quote message");

        log.info("Out testSbeQuote");
    }
}

File: frontend/tests/messages/Quote.test.js
import React from 'react';
import { render, act, fireEvent } from '@testing-library/react';
import WebSocketServer from 'jest-websocket-mock';
import { WebSocketProvider, useWebSocket } from 'src/contexts/WebSocketContext.js';

import QuoteEncoder from 'src/aeron/v1/QuoteEncoder.js'
import QuoteDecoder from 'src/aeron/v1/QuoteDecoder.js'
import MessageHeaderEncoder from 'src/aeron/MessageHeaderEncoder.js';

// Mock TextEncoder & TextDecoder
import TextEncoder from '../aeron/TextEncoder.js';
import TextDecoder from '../aeron/TextDecoder.js';
global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;

const data = {
    amount: { mantissa: 1000, exponent: -2 },
    currency: 'USD',
    fxRate: { mantissa: 123456, exponent: -5 },
    side: 'BUY',
    symbol: 'EURUSD',
    transactTime: '20240101-00:00:00.000',
    quoteRequestID: 'test-request',
    quoteID: 'test-quote-id',
    secondaryAmount: { mantissa: 100, exponent: -2 },
};

const TestComponent = () => {
    const { sendMessage } = useWebSocket();
    const sendTestMessage = () => {

        const buffer = new ArrayBuffer(QuoteEncoder.BLOCK_LENGTH + MessageHeaderEncoder.ENCODED_LENGTH);
        const headerEncoder = new MessageHeaderEncoder();
        const encoder = new QuoteEncoder();

        // Encode the quote
        encoder.wrapAndApplyHeader(buffer, 0, headerEncoder);
        encoder.encodeamount(data.amount);
        encoder.currency(data.currency);
        encoder.side(data.side);
        encoder.symbol(data.symbol);
        encoder.transactTime(data.transactTime);
        encoder.quoteID(data.quoteID);
        encoder.quoteRequestID(data.quoteRequestID);
        encoder.encodefxRate(data.fxRate);
        encoder.encodesecondaryAmount(data.secondaryAmount);

        // Send the message on the WebSocket
        sendMessage(buffer);
    };

    return <button onClick={sendTestMessage}>Send Quote</button>;
};

describe('WebSocket Quote integration test', () => {
    let server;

    beforeEach(() => {
        server = new WebSocketServer('ws://localhost:8294');
    });

    afterEach(() => {
        WebSocketServer.clean();
    });

    it('receives a Quote message via WebSocket', async () => {
        const { getByText } = render(
            <WebSocketProvider url="ws://localhost:8294">
                <TestComponent />
            </WebSocketProvider>
        );

        await server.connected;
        await act(async () => {
            fireEvent.click(getByText('Send Quote'));
        });

        // Allow server to process the message
        await new Promise((resolve) => setTimeout(resolve, 50));

        // Log all server messages
        console.log('Server messages after sending:', server.messages);
    
        // Verify the message was received on the server
        const messages = server.messages.filter(
            (message) => !(typeof message === 'string' && message.includes('Connection established'))
        );
        expect(messages).toHaveLength(1);
    
        // Received message
        const receivedMessage = messages[0];
        console.log('Received message:', receivedMessage);

        // Decode the received message
        const decoder = new QuoteDecoder();
        const buffer = new Uint8Array(receivedMessage).buffer;
        decoder.wrap(buffer, MessageHeaderEncoder.ENCODED_LENGTH);

        console.log('Decoding received message...');

        const decodedData = {
            amount: decoder.decodeamount(),
            currency: decoder.currency(),
            side: decoder.side().replace(/\0/g, ''),
            symbol: decoder.symbol(),
            transactTime: decoder.transactTime(),
            quoteRequestID: decoder.quoteRequestID().replace(/\0/g, ''),
            quoteID: decoder.quoteID().replace(/\0/g, ''),
            fxRate: decoder.decodefxRate(),
            secondaryAmount: decoder.decodesecondaryAmount()
        };

        // Assert the decoded data matches the original data
        expect(decodedData).toEqual(data);
    });
});